identity-core/src/common/one_or_set.rs

```diff
@@ -0,0 +1,484 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use core::fmt::Debug;
+use core::fmt::Formatter;
+use core::hash::Hash;
+use core::iter;
+use core::mem::replace;
+use core::ops::Deref;
+use core::slice::from_ref;
+
+use serde::de;
+use serde::Deserialize;
+use serde::Serialize;
+
+use identity_diff::Diff;
+use identity_diff::DiffVec;
+
+use crate::common::KeyComparable;
+use crate::common::OrderedSet;
+use crate::error::Error;
+use crate::error::Result;
+
+/// A generic container that stores exactly one or more unique instances of a given type.
+///
+/// Similar to [`OneOrMany`](crate::common::OneOrMany) except instances are guaranteed to be unique,
+/// and only immutable references are allowed.
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct OneOrSet<T>(OneOrSetInner<T>)
+where
+  T: KeyComparable;
+
+// Private to prevent creations of empty `Set` variants.
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
+#[serde(untagged)]
+enum OneOrSetInner<T>
+where
+  T: KeyComparable,
+{
+  /// A single instance of `T`.
+  One(T),
+  /// Multiple (one or more) unique instances of `T`.
+  #[serde(deserialize_with = "deserialize_non_empty_set")]
+  Set(OrderedSet<T>),
+}
+
+/// Deserializes an [`OrderedSet`] while enforcing that it is non-empty.
+fn deserialize_non_empty_set<'de, D, T: serde::Deserialize<'de> + KeyComparable>(
+  deserializer: D,
+) -> Result<OrderedSet<T>, D::Error>
+where
+  D: de::Deserializer<'de>,
+{
+  let set: OrderedSet<T> = OrderedSet::deserialize(deserializer)?;
+  if set.is_empty() {
+    return Err(de::Error::custom(Error::OneOrSetEmpty));
+  }
+
+  Ok(set)
+}
+
+impl<T> OneOrSet<T>
+where
+  T: KeyComparable,
+{
+  /// Constructs a new instance with a single item.
+  pub fn new_one(item: T) -> Self {
+    Self(OneOrSetInner::One(item))
+  }
+
+  /// Constructs a new instance from a set of unique items.
+  ///
+  /// Errors if the given set is empty.
+  pub fn new_set(set: OrderedSet<T>) -> Result<Self> {
```

Comment:
nit: 
These could be perhaps be private as one can use the `From` and `TryFrom` traits as constructors.

Reply:
I would prefer to keep these because they indicate the canonical ways of constructing a `OneOrSet` in absence of an exposed enum. `From` and `TryFrom` implementations are not obvious or easy to find and would not be the first place to look to instantiate a struct.

This is similar to why we have constructors like `IotaDocument::from_verification_method` and `IotaDocument::try_from_core` (which I also argue should be renamed `new_from_verification_method` and `new_from_core`) rather than `From` and `TryFrom` implementations. This is also a point for the Wasm bindings where we would need to add `From` implementations as functions on the struct itself, but that's moot in this case since `OneOrSet` will not be exposed directly.

Reply:
That is fair. I won't complain either way.